### PR TITLE
Introduce an experimental `python` pipeline operator

### PR DESCRIPTION
**Pull request by lava**
This experimental operator is intended for situations where development velocity is more important than performance; it allows the user to run arbitrary code against the (converted-to-python) contents of a table slice.

Example:
```
$ tenzir 'show build | python "import math; sanitizers.root = math.sqrt(2*len(tree_hash) + len(type))"'

{
  "type": "Debug",
  "tree_hash": "0a178da187a13feefa10c1f18f7f20e9",
  "assertions": true,
  "sanitizers": {
    "address": true,
    "undefined_behavior": false,
    "root": 8.306623862918075
  }
}
```

### Remaining Work
- [x] Make use of the `extra`s feature to avoid installing false dependencies for the operator venvs.
- [x] Put the venvs under a suitable cache location.
- [x] Synchronize venv creation to prevent write races.
- [x] Implement a "mutable pytenzir" config option.
- [x] Wire python errors into the operator API.
- [x] Extend the test coverage.
- [x] Clean up the commit history.
- [x] Improve the docs.

### CODE REVIEW

File: tenzir/functional-test/tests/setup_suite.bash [hunk 1]
```diff
@@ -16,9 +16,9 @@ export BATS_LIB_PATH=${BATS_LIB_PATH:+${BATS_LIB_PATH}:}${SCRIPT_DIR}/..
 BATS_SUITE_DIRNAME="${BATS_TEST_DIRNAME}"
 export BATS_SUITE_DIRNAME
 
-export PYTHONPATH=$PYTHONPATH${PYTHONPATH:+:}$SCRIPT_DIR/../../../python
+# Normalize the environment, but pass through the pytenzir operator helper.
+unset $(printenv | grep -v '^TENZIR_PLUGINS__PYTHON__IMPLICIT_REQUIREMENTS=' | grep -o '^TENZIR[^=]*' | paste -s -)
```
Comment (tobim) [inline tenzir/functional-test/tests/setup_suite.bash:20]: Consider doing this dependent on some IMPURE env variable.

File: libtenzir/src/format/arrow.cpp [hunk 2]
```diff
@@ -109,6 +111,7 @@ arrow_istream_wrapper::Read(int64_t nbytes) {
 }
 
 arrow_fd_wrapper::arrow_fd_wrapper(int fd) : fd_{fd} {
+  set_mode(::arrow::io::FileMode::READ);
```
Comment (tobim) [inline libtenzir/src/format/arrow.cpp:114]: Not sure if we really need this.

File: python/pytenzir/tools/python_operator_executor.py [hunk 3]
```diff
@@ -277,6 +277,9 @@ def main():
             batch_out = _execute_user_code(batch_in, code)
             writer = pa.ipc.RecordBatchStreamWriter(ostream, batch_out.schema)
             writer.write(batch_out)
+            # TODO: Consider adding writer.close() here and reading the closing
+            # null sequence on the c++ side. Experimental evidence suggests we
+            # don't need that but in principle we should.
```
Comment (tobim) [inline python/pytenzir/tools/python_operator_executor.py]: We should actually do this on both ends.

File: libtenzir/src/format/arrow.cpp (hunk 2 above)
Comment (lava) [reply libtenzir/src/format/arrow.cpp:114]: We noticed that `READ` is the default value for the mode anyways

File: tenzir.yaml.example [hunk 4]
```diff
@@ -18,6 +18,14 @@ tenzir:
   # The file system path used for persistent state.
   db-directory: "tenzir.db"
 
+  # The file system path used for persistent state.
+  # Defaults to one of the following paths, selecting the first that is
+  # available:
+  #   - $XDG_CACHE_HOME
+  #   - $XDG_HOME_DIR/.cache/tenzir or
```
Comment (dominiklohmann) [inline tenzir.yaml.example:25]: - If XDG_HOME_DIR is not set, you should probably use HOME instead.
- On macOS, caches should go under `$HOME/Library/Caches/tenzir` instead, unless either of the XDG variables are defined.

Comment (lava) [reply tenzir.yaml.example:25]: Are the `XDG_` variables usually defined on mac, or would that be the exception?

Comment (dominiklohmann) [reply tenzir.yaml.example:25]: They are undefined, some people define them because some software respects it, but it's far from standard behavior.

**Review by tobim — APPROVED** [review]
This has been a long time coming. Thanks to everyone for their input and help. Especially @lava and @jachris.

**Pull request merged**